### PR TITLE
Dataspaces

### PR DESCRIPTION
**Pull request by kshitij-v-mehta**

### CODE REVIEW

File: codar/cheetah/launchers.py [hunk 1]
```diff
@@ -165,9 +166,16 @@ def create_group_directory(self, campaign_name, group_name, runs,
                             method_opts = value_tokens[1]
 
                         adios_params.adios_xml_transport(
-                            xml_filepath, pv.group_name, method_name, method_opts)
+                            xml_filepath, pv.group_name, method_name,
+                            method_opts)
                     else:
-                        raise "Unrecognized adios param"
+                        raise Exception("Unrecognized adios param")
```
Comment (bd4) [inline codar/cheetah/launchers.py]: I want to start using CheetahException or a subclass for all exceptions

File: codar/cheetah/model.py [hunk 2]
```diff
@@ -548,6 +551,105 @@ def get_app_param_dict(self):
...
+        Add dataspaces support for this run.
+        Creates a new RC with dataspaces server as the exe.
+        :param rc_list: List of RC objects coupling using dataspaces
+        :param machine_name: Current machine
+        :return:
+        """
+
+        # Sanity check. rc list for coupling must have >1 RCs
+        if len(rc_list) == 1:
+            raise Exception("Atleast 2 codes needed for Dataspaces coupling. "
```
Comment (bd4) [inline codar/cheetah/model.py]: Should also be CheetahException

File: codar/cheetah/model.py [hunk 3]
```diff
@@ -548,6 +551,105 @@ def get_app_param_dict(self):
         (does not include nprocs or exe paths)."""
         return self.instance.as_dict()
 
+    def add_dataspaces_support(self, machine_name):
```
Comment (bd4) [inline codar/cheetah/model.py]: Isn't this the machine object, not the string name?

File: codar/cheetah/model.py [hunk 4]
```diff
@@ -548,6 +551,105 @@ def get_app_param_dict(self):
...
+        sleep_after = 0
+        for code in self.codes:
+            exe = self.codes[code]['exe']
+            if 'dataspaces_server' in exe:
+                ds_server = exe
+                ds_rc_name = code
+                sleep_after = self.codes[code].get('sleep_after', 0)
+
+        if not ds_server:
+            raise Exception("Dataspaces server needs to be specified in codes")
```
Comment (bd4) [inline codar/cheetah/model.py]: CheetahException

File: codar/cheetah/model.py [hunk 5]
```diff
@@ -548,6 +551,105 @@ def get_app_param_dict(self):
...
+
+        # Copy the configuration file dataspaces.conf
+        ds_conf = os.path.join(self.codes_path, "dataspaces.conf")
+        if not os.path.isfile(ds_conf):
+            raise Exception("Could not find dataspaces.conf in " +
```
Comment (bd4) [inline codar/cheetah/model.py]: CheetahException

File: codar/cheetah/launchers.py [hunk 6]
```diff
@@ -165,9 +166,16 @@ def create_group_directory(self, campaign_name, group_name, runs,
...
+
+                # Insert dataspaces server instances if RCs will couple
+                # using dataspaces.
+                # This must be called after the ADIOS params are parsed and
+                # the final ADIOS XML is generated
+                run.add_dataspaces_support(machine)
```
Comment (bd4) [inline codar/cheetah/launchers.py]: This is based on master and not dev branch right? There will likely be a conflict, with the auto-nodes feature some things were moved around. The important thing for the merge is to make sure that dataspaces support is added before the node calculation is done.

File: codar/cheetah/model.py [hunk 7]
```diff
@@ -548,6 +551,105 @@ def get_app_param_dict(self):
...
+                node_layout = d[ds_rc_name]
+        if node_layout is None:
+            node_layout = machine.dataspaces_servers_per_node
+
+        rc = RunComponent(rc_name, ds_server, args,
+                          nprocs=num_servers, sleep_after=sleep_after,
+                          working_dir=self.run_path)
+
+        self.node_layout.add_node({rc_name: node_layout})
+        self.run_components.insert(0, rc)
```
Comment (bd4) [inline codar/cheetah/model.py]: Which should be first, sosflow or dataspaces?

File: codar/cheetah/launchers.py (hunk 1 above)
Comment (kshitij-v-mehta) [reply codar/cheetah/launchers.py]: Ok, I will make the changes at all place where it is applicable

File: codar/cheetah/model.py (hunk 3 above)
Comment (kshitij-v-mehta) [reply codar/cheetah/model.py]: Thats right. I will modify.

File: codar/cheetah/launchers.py (hunk 6 above)
Comment (kshitij-v-mehta) [reply codar/cheetah/launchers.py]: This is based on dev but before the campaign status PR was merged.
> The important thing for the merge is to make sure that dataspaces support is added before the node calculation is done
Yep

File: codar/cheetah/model.py (hunk 7 above)
Comment (kshitij-v-mehta) [reply codar/cheetah/model.py]: Doesn't matter

Comment (kshitij-v-mehta) [reply codar/cheetah/model.py]: They run independently of each other, so it doesn't matter what order they are started in. SOSflow does not profile DataSpaces, and it also it not a DataSpaces client that DataSpaces needs to be aware of.

**Pull request merged**